Sources/MongoSwift/MongoClient.swift

```diff
@@ -548,8 +548,19 @@ public class MongoClient {
         using connection: Connection? = nil,
         session: ClientSession? = nil
     ) -> EventLoopFuture<T.OperationResult> {
+        guard !self.isClosed else {
+            return self.makeFailedFuture(MongoClient.ClosedClientError)
+        }
+        if case .ended = session?.state {
+            return self.makeFailedFuture(ClientSession.SessionInactiveError)
+        }
         return self.operationExecutor.execute(operation, using: connection, client: self, session: session)
     }
+
+    /// Creates an `EventLoopFuture<T>` that fails with the provided error.
+    internal func makeFailedFuture<T>(_ error: Error) -> EventLoopFuture<T> {
```

Comment:
so last time I fixed the bulk write error validation problem by just moving it into the operation's execute method, but I ran into a few more validation-related cases in this PR that were tricky to move over. I also realized that waiting to fail until we are in the thread pool is less optimal performance-wise if we can detect a problem earlier.

For that reason, I've added a `makeFailedFuture` method on `OperationExecutor` (along with this helper, for convenience) that allows us to generate failed futures without entering the thread pool.

Reply:
In my change streams work, I similarly need to access the underlying `execute` on the operation executor, meaning we'd have a helper for every method on `OperationExecutor` (except for `close`). This kinda makes the helpers feel redundant when we can just expose the executor itself as `internal` instead. What do you think?

Reply:
that's a good point. it also could lead to bugs if you forget the helper exists and just go directly to the executor since there are checks here that don't happen there. I'm fine with getting rid of the helper and just directly accessing the executor.

Reply:
I'll probably just do this in a separate PR right now since it is going to touch a lot of places unrelated to this PR